Complete AMBER --> OpenMM force-field conversion

rafwiewiora opened the pull request:
@peastman @jchodera @swails 

Done! I present to you the all-in-one solution to all your AMBER --> OpenMM-ffXML converting fantasies. 

Everything here, perhaps easier to see than in the diff: https://github.com/choderalab/openmm/tree/forcefield_conversion_bonanza/devtools/forcefield-scripts/amber

**Summary**: call `python amber2omm.py -v`. Requires the `files/` dir. Outputs `ffxml/` containing all converted XMLs and `log.csv` - log of the validation energies. 

**General**
Script supports command line arguments, and so `python amber2omm.py -h`:
```
usage: amber2omm.py [-h] [--input INPUT] [--input-format INPUT_FORMAT]
                    [--verbose] [--no-log] [--protein-test] [--nucleic-test]
                    [--protein-ua-test] [--phospho-protein-test] [--gaff-test]

AMBER --> OpenMM forcefield conversion script

optional arguments:
  -h, --help            show this help message and exit
  --input INPUT, -i INPUT
                        path of the input file. Default: "files/master.yaml"
  --input-format INPUT_FORMAT, -if INPUT_FORMAT
                        format of the input file: "yaml" or "leaprc". Default:
                        "yaml"
  --verbose, -v         turns verbosity on
  --no-log              turns logging of energies to log.csv off
  --protein-test        validate resulting XML through protein tests
  --nucleic-test        validate resulting XML through nucleic acid tests
  --protein-ua-test     validate resulting XML through united-atom protein
                        tests
  --phospho-protein-test
                        validate resulting XML through phosphorylated protein
                        tests
  --gaff-test           validate resulting XML through small-molecule (GAFF)
                        test
```

With the defaults as set, all you need to do is have the script and the `files/` directory and call `python amber2omm.py`. `-v` or `--no-log` as wanted. 

Output:
* `ffxml/` with all output XMLs
* `log.csv` - csv, machine readable log, containing AMBER, OpenMM, and relative (abs(AMBER-OpenMM) / AMBER) energies  - you can see what this looks like here: https://github.com/choderalab/openmm/blob/forcefield_conversion_bonanza/devtools/forcefield-scripts/amber/log.csv . This is overwritten if already present on a new run, use `--no-log` to disable. 
* (LeAP is called extensively by the script, and outputs its own `leap.log` too)

`-v` will turn on printing of the progress-tracking comments from the script. Warnings issued by ParmEd (warnings are by default filtered to error, but in some cases this had to be relaxed to 'always' because of the nature of the files converted) are printed always. The output of LeAP is always redirected to `dev/null`.

**YAML input**
By default the script takes a YAML file input, `files/master.yaml` has everything that is going on here. 
There's only a few rules to the required structure of the YAML and it will be very easily extendable to future forcefields. 

First entry in the YAML must be:
``` yaml
- sourcePackage: AmberTools
  sourcePackageVersion: 15
```

a MODE declaration follows:
``` yaml
- MODE: LEAPRC
```

There are two modes: LEAPRC and RECIPE. LEAPRC - self-explanatory, RECIPE - this is for water-ion conversions, RECIPE because your input files are a mix of `.dat`, `frcmod` and `.lib`, rather than leaprc. 

We're in LEAPRC mode - this is used for all protein/nucleic acids forcefield. So all the FFs follow, example:
``` yaml
- Source: leaprc.ff14SB
  Reference:
  - >-
    Maier, J.A., Martinez, C., Kasavajhala, K., Wickstrom, L., Hauser, K.E., and Simmerling, C. (2015).
    ff14SB: Improving the Accuracy of Protein Side Chain and Backbone Parameters from ff99SB. J. Chem. Theory Comput. 11, 3696-3713.
  Test:
  - protein
  - nucleic
```

I think that's self-explanatory, all fields required. 

There's an optional `Options` field, which allows changes to parameters if the default is no good for this case:
``` yaml
- Source: leaprc.phosaa10
  Reference:
  - >-
    Steinbrecher, T., Latzer, J., and Case, D.A. (2012). Revised AMBER parameters for bioorganic phosphates.
    J. Chem. Theory Comput. 8, 4405-4412.
  Options:
    filter_warnings: always
    write_unused: True
  Test:
  - protein_phospho
```

And so we do all protein / nucleic acid FFs. 

Then we reach water & ions. MODE is changed to RECIPE:
``` yaml
- MODE: RECIPE
```

and extra source package is declared (water models are converted manually - we just supply them as an ffxml in `files/` - I have `tip3p.xml`, `tip4pew.xml`, `spce.xml` from the current distribution of OpenMM with some changes to make them the 'newest' format (e.g. no classes used, only types). These ffxmls are integrated together with the converted ion parameters to make all the output. So we have OpenMM 7.0 listed as the source of these files - hence the extra input). 

``` yaml
- sourcePackage2: OpenMM
  sourcePackageVersion2: 7.0
```

Then again, we're going ffXML by ffXML - but we need a few extra fields. Two examples:
A 'standard' file - water model + JC monovalent ions + compromise set +2 ions
``` yaml
- Source:
  - parm/frcmod.ionsjc_tip3p
  - parm/frcmod.ionslrcm_cm_tip3p
  - lib/atomic_ions.lib
  Solvent_source: tip3p.xml
  Solvent: tip3p
  Name: tip3p_standard
  Reference:
  - >-
    Joung, I.S., and Cheatham, Thomas E. (2008).
    Determination of Alkali and Halide Monovalent Ion Parameters for Use in Explicitly Solvated Biomolecular Simulations. J. Phys. Chem. B 112, 9020-9041.
  - >-
    Joung, I.S., and Cheatham, T.E. (2009).
    Molecular Dynamics Simulations of the Dynamic and Energetic Properties of Alkali and Halide Ions Using Water-Model-Specific Ion Parameters. J. Phys. Chem. B 113, 13279-13290.
  - >-
    Li, P., Roberts, B.P., Chakravorty, D.K., and Merz, K.M. (2013).
    Rational Design of Particle Mesh Ewald Compatible Lennard-Jones Parameters for +2 Metal Cations in Explicit Solvent. J. Chem. Theory Comput. 9, 2733-2748.
  - >-
    Jorgensen, W.L., Chandrasekhar, J., Madura, J.D., Impey, R.W., and Klein, M.L. (1983).
    Comparison of simple potential functions for simulating liquid water. The Journal of Chemical Physics 79, 926-935.
  Test:
  - water_ion
```
An 'overloading' set: HFE +2, +3 and +4 ions for tip3p water. 

``` yaml
- Source:
  - parm/frcmod.ionslrcm_hfe_tip3p
  - parm/frcmod.ions34lsm_hfe_tip3p
  - lib/atomic_ions.lib
  Standard: tip3p_standard
  Solvent: tip3p
  Name: tip3p_HFE_multivalent
  Reference:
  - >-
    Li, P., Roberts, B.P., Chakravorty, D.K., and Merz, K.M. (2013).
    Rational Design of Particle Mesh Ewald Compatible Lennard-Jones Parameters for +2 Metal Cations in Explicit Solvent. J. Chem. Theory Comput. 9, 2733-2748.
  - >-
    Li, P., Song, L.F., and Merz, K.M. (2015).
    Parameterization of Highly Charged Metal Ions Using the 12-6-4 LJ-Type Nonbonded Model in Explicit Water. J. Phys. Chem. B 119, 883-895.
  Test:
  - water_ion
```

* Source - AMBER input files

* Solvent_source - the water file in `files/` for the standard (i.e. water model containing) XMLs **or** Standard - this is the same as the `Name` field for the appropriate standard (water model containing) XML - we need to know that, because for the 'overloading' sets both that XML and the standard XML need to be loaded for energy testing

* Solvent - this is the name of the solvent, this is necessary to avoid hardcoding of recognition of what solvent you're using from the names of the files etc. - and knowing which solvent you're using is necessary for energy validations. 

* Name - the desired name of the ffxml. (For proteins and nucleic this is done by the script, which a product of `leaprc.ff14SB'` will call `ff14SB.xml` etc.)

Should you want to provide a different YAML (shorter version, different completely, whatever you want), script will take it with:
```
python amber2omm.py --input name_of_your_yaml.yaml
```

The outputs of any YAML will be written to `ffxml/`. 

You can also provide a leaprc of your choosing via:
```
python amber2omm.py --input name_of_your_leaprc --input-format leaprc
```

The output of any leaprc will be written to the `./`. 

**A few remarks on the water models and ions**

As I've said before already, the system is:
* water models converted manually, i.e. taken existing files, put them in `files/`, script will merge them with appropriate ions

* we have *standard* sets: `tip3p_standard.xml`, `tip4pew_standard.xml`, `spce_standard.xml`: water model + JC monovalent ions + compromise set +2 ions

* for each water model we then have an HFE and IOD sets - +2/+3/+4 ions, all have templates set to `overload = "1"`. (`tip3p_HFE_multivalent.xml`, `tip3p_IOD_multivalent.xml` etc.)

* usage is to always load in a standard, and then you can overload +2's and add +3 and +4 with the HFE or IOD files

* naming of the water atom types has stayed as was (`tip3p-O`)

* naming of the ion atom types is `name_of_set (dash) amber_atom_type_name`, e.g. `tip3p_standard-Na+`, `tip3p_HFE_multivalent-Zn2+`. 


**TODO**

* Please look through, see what you like, what you don't like!

* https://github.com/pandegroup/openmm/pull/1373 must be merged before the script is functional.

* The impropers problem in nucleic acids is the only thing outstanding and I will attempt a solution next week. (https://github.com/choderalab/openmm/issues/9). Currently the improper energies are not validated for nucleic acid systems because of this. 

* @jchodera how are we doing on the Travis in this branch? Some testing appeared surprisingly some time ago, but I don't think it's testing the conversion - this has not been merged https://github.com/choderalab/openmm/pull/6. (Note the name of the script has changed now.)

* Add docstrings in the script, write a README.

peastman commented:
That's great!  It may take me a little while to go through all of this. :)

rafwiewiora commented:
Take your time! :)

swails commented on devtools/forcefield-scripts/amber/amber2omm.py:
```diff
@@ -0,0 +1,844 @@
+        if args.gaff_test:
+            validate_gaff(ffxml_name, args.input)
+    else:
+        raise RuntimeError('Wrong input_format chosen.')
```
I'd use `ValueError` here, or just `sys.exit('Wrong input_format chosen')`

rafwiewiora replied:
So I was using `RuntimeError` everywhere in the script (except for a few `AssertionErrors` in energy testing) - 'cause it means 'error while running the code' essentially to me. So are you saying only change in this specific place or is this gonna need such changes throughout?

swails replied:
To me `RuntimeError` is something that indicates some kind of internal inconsistency (same with `AssertionError`.  Since this is being done inside `main()` and isn't going to be caught, the end result of this statement is a hard system exit with a uselessly verbose traceback.  Handling this condition would be much cleaner by replacing it with a simple `sys.exit('Wrong input_format chosen')` (which still exits with a non-zero exit status, but doesn't clutter the output with a traceback).

Basically, IMO you should use `sys.exit` where you intend for an error to quit execution.  Raise an exception when you really want a Traceback.

rafwiewiora replied:
Ok, thanks! I'll go through and give all the Errors some flavor, it's fewer than I thought in there anyway reading through again. 

Re `AssertionError`: what kind of error would you throw in a `if x > tolerance` numerical test then?

swails commented on devtools/forcefield-scripts/amber/amber2omm.py:
```diff
@@ -0,0 +1,844 @@
+        if i[1] != 0:
+            rel_energies.append((i[0], (abs(i[1]-j[1])/abs(i[1]))))
+        else:
+            if j[1] != 0:
```
Change direct comparisons to 0 to be > small number.  In several of my tests with OpenMM/ParmEd/Amber I've seen energies ~1e-300, which would fail this test but clearly still be OK.

rafwiewiora replied:
Oh great call, thanks!

swails commented:
@rafwiewiora -- this is excellent work.  I made a few small notes.  I think this will be a valuable addition to OpenMM and the field in general.  Your hard work helped improve Amber as well by fixing some previously undetected issues with some of our force fields.

On a related note, I'm going to an Amber developer meeting in San Diego (right after the ACS) at the end of the week.  What would you think about me presenting some of this work (of course listing you and @jchodera as coauthors on the talk)?  I was going to talk mainly about ParmEd, anyway, and I think this would make a great success story.

rafwiewiora commented:
Thanks a lot @swails, you've been of super great help throughout! And I think I learnt a lot about good quality coding, most of it from reading ParmEd... :D 

> On a related note, I'm going to an Amber developer meeting in San Diego (right after the ACS) at the end of the week. What would you think about me presenting some of this work (of course listing you and @jchodera as coauthors on the talk)? I was going to talk mainly about ParmEd, anyway, and I think this would make a great success story.

I think that sounds awesome, yes please, as long as @jchodera is ok with that? Would you like me to prepare something for you (better docs, slides?)?

swails commented:
A slide would be awesome.  It's a short talk, and I just wanted to make the Amber community aware of the work that's gone into porting the Amber force fields as well as the efforts made in validation (and throw in some stuff we learned about Amber during the process).

rafwiewiora commented:
Ok, cool, I'll work on it Monday at the latest!

rafwiewiora commented:
I wonder what you guys think about the relative AMBER - OpenMM calculation and validation. 

One of the things that @swails pointed out in the line comments yesterday is to turn all comparisons of numbers to 0 into comparisons to a very small number (side question is what such small number to use? I see you're using 1e-8 as TINY in ParmEd - but is this so high because that's the tolerance you give for example for LJ scaling factors? I was thinking - on double precision, the smallest float should be something like ~1e-308 I think, so go with `abs(x) =< 1e-300` for zero comparisons?) 

But then I thought, well if we had energies of 1e-300 and -1e-300, so difference 2e-300 and relative difference 2e-300/1e-300 = 2 - I'm failing the testing very badly, but clearly these are very close to zero so I should say this is okay. And this website talks about this more: http://floating-point-gui.de/errors/comparison/ So how do I protect myself against this kind of situation? 

Until I added the logging functionality I was doing this validation with `numpy.testing.assert_allclose` and, without looking into source, I presume that is well-protected against these kinds of problems. I stopped using it and replaced with my own calculation because I needed to log the relative energies for printing in the log and didn't want to make this more expensive by calculating the same things twice. But now I wonder whether I should make that sacrifice and do my own calculation for logging, but use `numpy` for testing and be safe against these close-to-zero cases?

EDIT: Actually, `numpy.testing` is not protected against this problem, just tried it. So I'm just going to leave everything as it is, but I'm going to set TINY to 1e-10 and compare to that instead of zero. If we ever have a freak case where we have one energy of 1e-9 and another of -1e-9, which would fail this test - we can decide at that time whether that should still pass or not. I think 1e-10 is reasonable for now. Does that sound good?

EDIT2: @jchodera I've added `test/test.py` with a short test of leaprc and yaml conversion that you can hook up to Travis.

rafwiewiora commented:
Also a questions to @peastman: if we made the default `extraParticleIdentifier` `EP` in https://github.com/pandegroup/openmm/pull/1302, should the `tip4pew.xml` have `EP` in the element fields for `M`? I'm assuming right now the `M` stays as element `None` just because it cannot match with any element by name?

rafwiewiora commented:
(@peastman I wrote a question to you here before that I realized made no sense the moment I posted it, so please ignore if you get an email)

peastman commented:
> I was thinking - on double precision, the smallest float should be something like ~1e-308 I think, so go with abs(x) =< 1e-300 for zero comparisons?

In double precision, machine epsilon is about 1e-16, so that's the smallest relative tolerance that would have any meaning.  Smaller than that, and you're just saying the numbers have to be identically equal.  But you can't actually expect your results to be that close: the values you're comparing are the result of many calculations, and every intermediate calculation has the potential to add a bit more error.  Also, the energy is the sum of many terms with widely varying magnitude, and whose signs can be either positive or negative.  You can lose a lot of precision there.

rafwiewiora commented:
> In double precision, machine epsilon is about 1e-16, so that's the smallest relative tolerance that would have any meaning. Smaller than that, and you're just saying the numbers have to be identically equal. But you can't actually expect your results to be that close: the values you're comparing are the result of many calculations, and every intermediate calculation has the potential to add a bit more error. Also, the energy is the sum of many terms with widely varying magnitude, and whose signs can be either positive or negative. You can lose a lot of precision there.

Oh that's great to know! Well - I have the cutoff's set to 1e-5 for everything except impropers and 1e-2 for impropers - which I chose really from seeing what the relative differences were in an example where I was confident everything was fine code-wise. So that's exactly as you're saying - we're not gonna get as low as the machine epsilon here.

But my question was more about comparing an absolute energy value to zero here https://github.com/choderalab/openmm/blob/forcefield_conversion_bonanza/devtools/forcefield-scripts/amber/amber2omm.py#L395 - so that there's no division by zero in next line. So I've now gone with - NEARLYZERO is 1e-10, if AMBER energy is =< NEARLYZERO - don't calculate the relative difference, instead check that OpenMM energy is also =< NEARLYZERO. If not - exception, if yes - say relative difference is zero. (Just realized I need `abs(i[1])` and `abs(j[1])` there). So really this is just setting a threshold for approximating something as zero and I guess it's a compromise between:
- how big a number you're willing to call a zero
- how small a number has to get to cause some trouble with the 'being on both sides of zero' issue (this probably has some nice actual name) - e.g. with the current code (plus abs change) if I had AMBER energy - `1e-10`, OpenMM energy - `-1e-10` - relative difference will be set to zero. But if I had `1e-9` and `-1e-9`, the rel. difference is now 2 - exception. 

So the question here is: **what should the value of NEARLYZERO be, so that if I'm comparing two very small energies, but on opposite sides of zero, the exception means there's really a problem, rather than they are still very small numbers I would look at and say 'that's fine', but the division by the very small number makes the relative difference high?**

Does that make sense? (tell me if I'm very wrong in this whole thinking please)

peastman commented:
There's no universal answer to that.  It depends on what sort of values you're comparing, and what the expected errors in them are.  When you're computing the sum of a bunch of values, the expected absolute error scales as O(sqrt(n)) in the number of values you're adding.  If the values all have the same sign, then the relative error will scale in the same way.  But if they have varying signs, the relative error can be much larger than the absolute error.

Think of adding 123.456 and -123.345, each accurate to six significant figures.  The sum is 0.111.  Its absolute error is on the order of 0.001, the same as the two numbers you added.  But it's only accurate to three significant figures, so the relative error is 1000 times larger than in the two numbers you added.

jchodera commented:
Just catching up here. This is fantastic!

I'm 100% OK with @swails presenting anything here.  (Our work is always open by default.)

Reading through in detail, but first thoughts:
- Can we avoid using `omm` for "OpenMM" in general?  This is a branding issue.  The script should probably be called `amber2openmm` or `amber2ffxml` or something.
- A `README.md` based on your nice introduction in this thread would be great---even if just a dump of what you wrote here.

rafwiewiora commented:
> Can we avoid using omm for "OpenMM" in general? This is a branding issue. The script should probably be called amber2openmm or amber2ffxml or something.
> A README.md based on your nice introduction in this thread would be great---even if just a dump of what you wrote here.

Ha, I like that this is a branding issue! I went with `omm` because I think I saw it so many times in every internal function name in the `openmm/parameters.py` in ParmEd. Changing to openmm (I think that's much better for the brand then just ffxml, right?)

And will work on a README.md later today for sure!

rafwiewiora commented:
> There's no universal answer to that. It depends on what sort of values you're comparing, and what the expected errors in them are. When you're computing the sum of a bunch of values, the expected absolute error scales as O(sqrt(n)) in the number of values you're adding. If the values all have the same sign, then the relative error will scale in the same way. But if they have varying signs, the relative error can be much larger than the absolute error.
> 
> Think of adding 123.456 and -123.345, each accurate to six significant figures. The sum is 0.111. Its absolute error is on the order of 0.001, the same as the two numbers you added. But it's only accurate to three significant figures, so the relative error is 1000 times larger than in the two numbers you added.

That's a very cool explanation @peastman, thanks so much! Ok, I will stay with `1e-10`, it seems reasonable intuitively, and it's got to be rather unlikely that you would get such a small number energy that would cause a problem here anyway. If anytime problems show up, we'll revisit.

rafwiewiora commented:
A thought about the overloading of ions: 
- problem: some _standard_ +2 ions do not have an equivalent in HFE or IOD +2 ions - i.e. when I load in `ForceField('tip3p_standard.xml', 'tip3p_HFE_multivalent.xml')` overloading will only 'do the job' for those +2 residues, which appear in both standard and overloading sets, we will be left with some _standard_ +2 ion templates available. 
- does a user expect that any +2 ion that will be available to them after such ForceField call are from appropriate HFE/IOD set, not the _standard_?
- could this lead to referencing wrong parameters?
- Should we not allow this, by including empty residue templates with the names of the problematic ones in the HFE/IOD sets, so that the old +2 templates are removed, but non-functional new templates are introduced?

swails commented:
Do you have a list of the residues that have no counterpart?

rafwiewiora commented:
Let me generate one.

swails commented:
Any chance you have a slide or two with porting/validation results I can present, @rafwiewiora?

rafwiewiora commented:
Sorry for the delay @swails, got hooked on MSMs and FAH data for the last 2 days. I will be putting something together for you tonight, you will have it by tomorrow morning for sure.

rafwiewiora commented:
@swails emailed you the slides. Also attaching here if anyone wants to check out. 
[amber2openmmslides.pdf](https://github.com/choderalab/openmm/files/177096/amber2openmmslides.pdf)

rafwiewiora commented:
@swails going back to the ions, looks like the templates included don't differ between solvents, and all HFE sets cover all standard ions, while all IOD sets miss these:

```
{'Ag', 'EU', 'PB', 'PD', 'PT', 'Ra', 'Sm', 'YB2'}
```

Not sure how often people might want to use these elements - I presume not very often. So this might be a (very) rare problem really.
